ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -116,41 +116,42 @@
-              // call. Therefore, when the call to add a listener returns, the initial notification will have been
-              // received and handled.
-              manager.addInstanceConfigChangeListener(clusterChangeHandler);
-              logger.info("Registered instance config change listeners for Helix manager at {}", zkConnectStr);
-              // Now register listeners to get notified on live instance change in every datacenter.
-              manager.addLiveInstanceChangeListener(clusterChangeHandler);
-              logger.info("Registered live instance change listeners for Helix manager at {}", zkConnectStr);
-            } catch (Exception e) {
-              initializationException.compareAndSet(null, e);
-            } finally {
-              initializationAttemptComplete.countDown();
+        // Initialize from every remote datacenter in a separate thread to speed things up.
+        Utils.newThread(() -> {
+          try {
+            HelixManager manager;
+            if (dcName.equals(clusterMapConfig.clusterMapDatacenterName)) {
+              manager = localManager;
+            } else {
+              manager =
+                  helixFactory.getZKHelixManager(clusterName, selfInstanceName, InstanceType.SPECTATOR, zkConnectStr);
+              logger.info("Connecting to Helix manager at {}", zkConnectStr);
+              manager.connect();
+              logger.info("Established connection to Helix manager at {}", zkConnectStr);
             }
+            DcInfo dcInfo = new DcInfo(dcName, entry.getValue(), manager, clusterChangeHandler);
+            dcToDcZkInfo.put(dcName, dcInfo);
+            dcIdToDcName.put(dcInfo.dcZkInfo.getDcId(), dcName);
+
+            // The initial instance config change notification is required to populate the static cluster
+            // information, and only after that is complete do we want the live instance change notification to
+            // come in. We do not need to do anything extra to ensure this, however, since Helix provides the initial
+            // notification for a change from within the same thread that adds the listener, in the context of the add
+            // call. Therefore, when the call to add a listener returns, the initial notification will have been
+            // received and handled.
+            manager.addInstanceConfigChangeListener(clusterChangeHandler);
+            logger.info("Registered instance config change listeners for Helix manager at {}", zkConnectStr);
+            // Now register listeners to get notified on live instance change in every datacenter.
+            manager.addLiveInstanceChangeListener(clusterChangeHandler);
+            logger.info("Registered live instance change listeners for Helix manager at {}", zkConnectStr);
+            if (!clusterMapConfig.clustermapListenCrossColo && manager != localManager) {
+              manager.disconnect();
```

Comment:
Wouldn't it be more efficient to move the property check up and skip the connection if we don't need it?

Reply:
We need to connect and get the cluster data from cross-colo ZK servers for initialization. We then close the connection, to stop listening for changes.